Attempt at fixing size mismatch in _compress_etc

MillionOstrich opened the pull request:
Fixes #12710 

This fixes the target_size being larger than the size calculated in `Image::create` caused when mmc is 1. 
I added the crash check because using the `mmc > 1 ` test rather than `mmc > 0` would memcpy out of bounds into `dst_data` if `mmc = 1` (which is the case in the test case described in that issue).

This does need another set of eyes though, as whilst this does stop the error I am not sure if this is the best or correct way to do so (e.g. do we actually want the mipmap if mmc is 1 or do we want to discard it?).

Commit:
Fix size mismatch in _compress_etc
Use j for inner loop

akien-mga commented:
CC @reduz and maybe @tagcup for review?

tagcup commented:
mmc == 1 when there are no mipmaps if memory serves, so mmc > 1 should be OK

tagcup commented on modules/etc/image_etc.cpp:
```diff
@@ -173,15 +174,17 @@ static void _compress_etc(Image *p_img, float p_lossy_quality, bool force_etc1_f
 	print_line("time encoding: " + rtos(OS::get_singleton()->get_ticks_msec() - t));
 
-	p_img->create(imgw, imgh, mmc > 1 ? true : false, etc_format, dst_data);
+	p_img->create(imgw, imgh, uses_mipmaps, etc_format, dst_data);
```
Changing only this line to
```cpp
p_img->create(imgw, imgh, p_img->has_mipmaps(), etc_format, dst_data);
```
should work, no?

tagcup commented on modules/etc/image_etc.cpp:
```diff
@@ -128,8 +128,9 @@ static void _compress_etc(Image *p_img, float p_lossy_quality, bool force_etc1_f
 	PoolVector<uint8_t>::Read r = img->get_data().read();
 
-	int target_size = Image::get_image_data_size(imgw, imgh, etc_format, p_img->has_mipmaps() ? -1 : 0);
 	int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 0;
```
Ah wait, this should probably be replaced with
```cpp
int mmc = (p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 0) + 1;
```

akien-mga replied:
Does `mmc` stand for "mipmaps count"? If so I guess it should just be this, no?
```cpp
int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 1;
```

tagcup replied:
No there was little bit of inconsistency about the mipmap count, which is the source of confusion. Check it but if memory serves, get_image_required_mipmaps probably returns the number of mipmaps without counting the image with original size.

tagcup replied:
If only the cpp code had some bit of documentation

tagcup replied:
like at least what does this non-trivial function do

tagcup replied:
BTW I wanted to check this with a recent build but my console is being flooded with

```
WARNING: not found: display/window/size/width
WARNING: not found: display/window/size/height
```

Is it just me or is this a known issue?

akien-mga replied:
Likely a regression from d09160a8b67fdc60e8108962c4e9bd4c0bc7f13e

MillionOstrich replied:
on second look would calling `int mmc = p_img->get_mipmap_count() + 1;` be ok here?

tagcup replied:
Just checked, get_image_required_mipmaps does count the original size image as well, so it should indeed be
```cpp
int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 1;
```

or simply

```cpp
int mmc = Image::get_image_required_mipmaps(imgw, imgh, etc_format);
```

(given the work etc needs to do, that "optimization" isn't worth it anyway)

tagcup replied:
@MillionOstrich there shouldn't be a +1 because get_mipmap_count already counts the original image.
Also, the required # of mipmaps depends on the format, so the # of mipmaps of the original and etc may not match

tagcup replied:
5cb1d06 fixed the "display/window/size/width" warning BTW

MillionOstrich replied:
@tagcup does that mean we want to remove the +1 from the for loop condition too?
`for (int i = 0; i < mmc + 1; i++) {`

tagcup replied:
looks that way
btw can you change the loop variable to j in the inner loop?

tagcup replied:
can you also check `modules/squish/image_compress_squish.cpp`? the code there should be more or less the same there so if there is a problem here, squish code would also need to be fixed

MillionOstrich replied:
I had a search around the codebase, `image_compress_squish` is using <= in the for loop for both compress and decompress. 

I also had a search for get_mipmap_count() as that calls the same method underneath -
 everything seems to be using +1:
```core\image.cpp(309): ERR_FAIL_INDEX_V(p_mipmap, get_mipmap_count() + 1, -1);
drivers\gles3\rasterizer_storage_gles3.cpp(772): int mipmaps = ((texture->flags & VS::TEXTURE_FLAG_MIPMAPS) && img->has_mipmaps()) ? img->get_mipmap_count() + 1 : 1;
editor\import\resource_importer_texture.cpp(245): int mmc = image->get_mipmap_count() + 1;
editor\import\resource_importer_texture.cpp(274): int mmc = image->get_mipmap_count() + 1;
modules\pvr\texture_loader_pvr.cpp(214): for (int i = 0; i <= new_img->get_mipmap_count(); i++) {
modules\squish\image_compress_squish.cpp(49): int mm_count = p_image->get_mipmap_count();
//^ uses <= in the loop later on
```

Results for `get_image_required_mipmaps`:
```modules\etc\image_etc.cpp(132): int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 1;
modules\squish\image_compress_squish.cpp(151): int mm_count = p_image->has_mipmaps() ? Image::get_image_required_mipmaps(w, h, target_format) : 0;
scene\resources\texture.cpp(591): int mipmaps = Image::get_image_required_mipmaps(tw, th, format);
```

The only case i can't see a +1 is in `texture.cpp:  StreamTexture::_load_data`, and that also has a >1 check too - but i don't understand what is happening in that method.

tagcup replied:
If you can confirm that it is indeed the case, the only change that is necessary would be

```cpp
p_img->create(imgw, imgh, p_img->has_mipmaps(), etc_format, dst_data);
```

tagcup replied:
or if you'd like to wait, I can give a proper check during weekend

akien-mga commented:
So what's the status, do we still want to make changes or is it good as is?

tagcup commented:
Sorry been a little busy recently
Based on the way other code is using the mipmap count, only changing mmc > 1 to mmc > 0 ir img->has_mipmaps should work, no? Or we can add 1 to where mmc is defined and adjust the rest accordingly

akien-mga commented:
@reduz Can you have a look at the proposed change and discussion?

MillionOstrich commented:
Updated this PR based on feedback.

- added 1 to mmc and changed loop condition to match
- use Image::has_mipmaps() to fix the call into Image::create() rather than inspecting mmc
- changed inner loop variable from i to j
